protocol/xmrmaker/instance_test.go

```diff
@@ -0,0 +1,185 @@
+	pswap "github.com/athanorlabs/atomic-swap/protocol/swap"
+	"github.com/athanorlabs/atomic-swap/protocol/xmrmaker/offers"
+	"github.com/athanorlabs/atomic-swap/tests"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	testWallet = "test-wallet"
+)
+
+type mockNet struct {
+	msgMu sync.Mutex  // lock needed, as SendSwapMessage is called async from timeout handlers
+	msg   net.Message // last value passed to SendSwapMessage
+}
+
+func (n *mockNet) LastSentMessage() net.Message {
+	n.msgMu.Lock()
+	defer n.msgMu.Unlock()
+	return n.msg
+}
+
+func (n *mockNet) SendSwapMessage(msg net.Message, _ types.Hash) error {
+	n.msgMu.Lock()
+	defer n.msgMu.Unlock()
+	n.msg = msg
+	return nil
+}
+
+func (n *mockNet) CloseProtocolStream(_ types.Hash) {}
+
+func newSwapManager(t *testing.T) pswap.Manager {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	db := pswap.NewMockDatabase(ctrl)
+	db.EXPECT().GetAllSwaps()
+	db.EXPECT().PutSwap(gomock.Any()).AnyTimes()
+
+	sm, err := pswap.NewManager(db)
+	require.NoError(t, err)
+	return sm
+}
+
+func newTestBackend(t *testing.T) backend.Backend {
+	pk := tests.GetMakerTestKey(t)
+	ec, chainID := tests.NewEthClient(t)
+
+	txOpts, err := bind.NewKeyedTransactorWithChainID(pk, chainID)
+	require.NoError(t, err)
+
+	var forwarderAddress ethcommon.Address
+	_, tx, contract, err := contracts.DeploySwapFactory(txOpts, ec, forwarderAddress)
+	require.NoError(t, err)
+
+	addr, err := bind.WaitDeployed(context.Background(), ec, tx)
+	require.NoError(t, err)
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	rdb := backend.NewMockRecoveryDB(ctrl)
+	rdb.EXPECT().PutContractSwapInfo(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	rdb.EXPECT().PutSwapPrivateKey(gomock.Any(), gomock.Any(), common.Development).Return(nil).AnyTimes()
+	rdb.EXPECT().PutSharedSwapPrivateKey(gomock.Any(), gomock.Any(), common.Development).Return(nil).AnyTimes()
+	rdb.EXPECT().PutMoneroStartHeight(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	rdb.EXPECT().GetMoneroStartHeight(gomock.Any()).Return(uint64(1), nil).AnyTimes()
+	rdb.EXPECT().PutSwapRelayerInfo(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+
+	extendedEC, err := extethclient.NewEthClient(context.Background(), ec, pk)
+	require.NoError(t, err)
+
+	bcfg := &backend.Config{
+		Ctx:                 context.Background(),
+		MoneroClient:        monero.CreateWalletClient(t),
+		EthereumClient:      extendedEC,
+		Environment:         common.Development,
+		SwapContract:        contract,
+		SwapContractAddress: addr,
+		SwapManager:         newSwapManager(t),
+		Net:                 new(mockNet),
+		RecoveryDB:          rdb,
+	}
+
+	b, err := backend.NewBackend(bcfg)
+	require.NoError(t, err)
+
+	return b
+}
+
+func newTestInstanceAndDB(t *testing.T) (*Instance, *offers.MockDatabase) {
```

Comment:
this function and all helper funcs above were copy pasted from swap_state_test.go